# Ejercicios - Clase 3

Ejercicios de los bloques de introducción a python. Listas, tuplas y diccionarios.

**Versión**: 1.0

**Fecha**: 13/10/2023

## Ejercicio 1

Vamos a calcular la `distancia euclidea` de dos puntos `p` y `q` en un espacio de tres dimensiones `D=3` definidos como tuplas. Recordad que la distancia euclidea sigue la ecuación:

$$d = \sqrt{\sum_{i=1}^{D=3} (p_i - q_i)^2}$$

In [5]:
import math

p = (5.4, -2.1, 3.2)
q = (3.2, 0.1, 0.9)

# TO-DO rellena esta celda calculando la distancia euclidea
s1 = (p[0] - q[0])**2
s2 = (p[1] - q[1])**2
s3 = (p[2] - q[2])**2
suma = s1 + s2 + s3
distancia = math.sqrt(suma)

In [6]:
print("La distancia euclidea entre {} y {} es {:.6f}".format(p, q, distancia))

La distancia euclidea entre (5.4, -2.1, 3.2) y (3.2, 0.1, 0.9) es 3.869108


Respuesta esperada:

- La distancia euclidea entre (5.4, -2.1, 3.2) y (3.2, 0.1, 0.9) es 3.869108

## Ejercicio 2

Podemos considerar una matriz como una lista de listas. Aquí tenéis un ejemplo para crear la siguiente matriz:

\begin{bmatrix}
    1 & -3 & 2\\
    2 & 3 & 1\\
 \end{bmatrix}

In [14]:
fila1 = [1, -3, 2]
fila2 = [2, 3, 1]
matriz = [fila1, fila2]
matriz

[[1, -3, 2], [2, 3, 1]]

El ejercicio consiste en crear una matriz con los números impares en la primera fila y los pares en la segunda fila hasta un número arbitrario N, como en este ejemplo:

\begin{bmatrix}
    1 & 3 & 5 & 7 & 9 & ... & 17 & 19\\
    2 & 4 & 6 & 8 & 10 & ... & 18 & 20\\
 \end{bmatrix}

In [38]:
N = 45

# TO-DO crea la matriz
fila_impares = list(range(1, N+1, 2))
fila_pares = list(range(2, N+1, 2))
matriz = [fila_impares, fila_pares]

print("Con N = {}, la matriz generada es la siguiente:".format(N))
print("  ", matriz[0])
print("  ", matriz[1])

Con N = 45, la matriz generada es la siguiente:
   [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45]
   [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44]


Posibles respuestas:

- Con N = 20, la matriz generada es la siguiente:

   [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
   
   [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

- Con N = 45, la matriz generada es la siguiente:

   [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45]
   
   [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44]

## Ejercicio 3

Vamos a leer un string que contiene una fecha con formato `Y-M-D`. Sin embargo, queremos imprimir el mes escrito con palabras con el siguiente formato `D de M de Y`. Utiliza un diccionario para ello:

In [42]:
fecha = "2023-09-30"

# TO-DO incluye aquí tu programa
year_str, month_str, day_str = fecha.split("-")
year, month, day = int(year_str), int(month_str), int(day_str)

month2string = {}
month2string[1] = "enero"
month2string[2] = "febrero"
month2string[3] = "marzo"
month2string[4] = "abril"
month2string[5] = "mayo"
month2string[6] = "junio"
month2string[7] = "julio"
month2string[8] = "agosto"
month2string[9] = "septiembre"
month2string[10] = "octubre"
month2string[11] = "noviembre"
month2string[12] = "diciembre"

fecha_formateada = "{} de {} de {}".format(day, month2string[month], year)

print("Hoy es " + fecha_formateada)

Hoy es 30 de septiembre de 2023


Posibles respuestas:

- Con input "2023-04-12":

  Hoy es 12 de abril de 2023

- Con input "2023-09-30":

  Hoy es 30 de septiembre de 2023

## Ejercicio 4

Vamos a crear un generador de nombres aleatorio a partir de dos listas y la librería RANDOM.

In [53]:
nombres = ["Christian", "Mateo", "Sofía", "Alejandro", "Isabel", "Rocío", "Santiago", "Valentín", "Pedro", 
           "Adriana", "Aitana", "David", "Ana", "Andrea", "Eva", "Iván", "Julián", "Alfonso", "Salvador",
           "Jesús", "Lucía", "Marta", "Gonzalo"]
apellidos = ["Oliva", "García", "Rodríguez", "González", "López", "Martínez", "Pérez", "Guijarro", "Santos",
             "Moya", "Lorenzo", "Benítez", "Ramos", "Caballero", "de los Monteros", "Rajoy", "Carrasco",
             "Molina", "Delgado", "Suarez"]

# TO-DO incluye aquí tu programa
nombre = random.choice(nombres)
apellido1 = random.choice(apellidos)
apellido2 = random.choice(apellidos)

print("Nombre generado: {} {} {}".format(nombre, apellido1, apellido2))

Nombre generado: Christian Lorenzo Lorenzo


Posibles respuestas:

- Nombre generado: Julián Rajoy Guijarro

- Nombre generado: Rocío Ramos Pérez

- Nombre generado: Ana Carrasco Oliva

- Nombre generado: Aitana Guijarro Santos

- Nombre generado: Marta Benítez Caballero

- Nombre generado: Christian Lorenzo Lorenzo